Increase compatibility with external cmake

RcColes opened the pull request:
Allow control of targets by external variables, and add include
directories to targets.

Signed-off-by: Raef Coles <raef.coles@arm.com>

## Description
Alter some of the cmake build system in order to allow better control of it by external projects. This allows, for example, a project to include two copies of mbedtls in its cmake (via include_directories) without causing target name clashes.

There shouldn't be any behavior changes in the cmake by default.

## Status
**READY**

## Requires Backporting

NO 

## Migrations

NO

## Additional comments

gilles-peskine-arm commented:
Thanks! There are other pull requests in flight that change the cmakefiles, in particular https://github.com/ARMmbed/mbedtls/pull/3337 and https://github.com/ARMmbed/mbedtls/pull/3159 . Have you checked for redundancy and conflicts?

gilles-peskine-arm commented:
@ronald-cron-arm As the de facto cmake expert, I'm assigning this to you. Please unassign if you don't have time to shepherd this PR.

ronald-cron-arm commented:
Yes I can handle this, no problem. This is related to things I've been looking at recently.

RcColes commented:
> Thanks! There are other pull requests in flight that change the cmakefiles, in particular #3337 and #3159 . Have you checked for redundancy and conflicts?

As best I can tell there is very little overlap. I think #3337 moves a a couple of `add_subdirectory` statements that we also move, but their placement seems like it'd work fine for us. #3159 uncapitalizes an `add_custom_target` that we also touch but that's functionally equivalent.

ronald-cron-arm commented on CMakeLists.txt:
```diff
@@ -205,22 +206,23 @@ if(ENABLE_ZLIB_SUPPORT)
-ADD_CUSTOM_TARGET(apidoc
-    COMMAND doxygen mbedtls.doxyfile
-    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/doxygen)
+if(ENABLE_DOCS)
```
This doc enablement seems to be a change on its own. Could you move that to its own commit?

RcColes replied:
Done. I've split it into 3 commits - I've gotten into bad habits of not properly partitioning commits

ronald-cron-arm replied:
I can see only 2 commits of yours: "Make api documentation target optional" and "Allow renaming of cmake targets". Otherwise, could you rebase on top of development? Commits are appearing in the commit list of this PR that are not related to it.

RcColes replied:
I think I've sorted it now, I removed one commit since I think it's not a change that should be upstreamed in it's current form / wasn't related to the other commits

RcColes commented:
> Could you explain why do you need those changes? What are your use cases? What are your issues with the way it is without this PR? We will have to maintain the changes thus we need to understand them and their purpose. In the PR description you mention two copies of mbedtls under the same cmake build system it seems, why do you need that?

The TF-M build system builds two versions of mbedtls(/mbedcrypto). One is a lightweight version that only enables functionality needed by the bootloader to save space, and the second is a full version is used by the crypto service to implement the PSA crypto api.

We're trying to alter the cmake to be more intuitive, and one part of this is that we've switched to building mbedtls by including the directory via `include_directories`. However, if you do this twice you run into errors where targets are multiply defined, which is a cmake error.

The idea behind these changes is that it allows all targets that would be multiply defined to either be disabled (such as the docs) or renamed by the including project, which removes the issue with multiple definitions.

ronald-cron-arm commented:
> We're trying to alter the cmake to be more intuitive, and one part of this is that we've switched to building mbedtls by including the directory via `include_directories`. However, if you do this twice you run into errors where targets are multiply defined, which is a cmake error.
> 
`include_directories` is a command to define include path. Do you mean `add_subdirectory`?

ronald-cron-arm commented:
> The TF-M build system builds two versions of mbedtls(/mbedcrypto). One is a lightweight version that only enables functionality needed by the bootloader to save space, and the second is a full version is used by the crypto service to implement the PSA crypto api.
> 
A quite natural way to build the same project two different ways with cmake is to define two different build trees. Did you consider this option?

RcColes commented:
> > The TF-M build system builds two versions of mbedtls(/mbedcrypto). One is a lightweight version that only enables functionality needed by the bootloader to save space, and the second is a full version is used by the crypto service to implement the PSA crypto api.
> 
> A quite natural way to build the same project two different ways with cmake is to define two different build trees. Did you consider this option?

That's what we are doing - I think there might be some confusion as to where the conflicts exist. If you have

```
add_subdirectory(mbedtls ${BUILD_DIR_1}
add_subdirectory(mbedtls ${BUILD_DIR_2}
```

This would do two builds in two different trees - you're correct in that this is the natural (and only?) way to build the same module twice in the same cmake.

However both of those include directories will run
```
add_library(mbedtls_static STATIC)
```
One for each time the mbedtls cmakelists is included. This will fail on the second run because cmake doesn't allow two targets to exist with the same name, even if they have been specified with different build directories. As far as I can tell cmake always treats targets as global, and has no concept of scoping of targets.

ronald-cron-arm commented:
Two build trees, yes, but I am not sure about the two add_subdirectory() to generate those two build trees. Adding twice the same sub-directory does not smell good to me. Did you consider other solutions, like copy of the build tree in other directory and run make from there, add_custom_target command ... ?

RcColes commented:
See I'd lean the other way. I see no reason that using add_subdirectory twice would be an issue - It even has some implicit support in that cmake allows you to select the build tree directory for a source tree dir, without which it would clearly be a bad idea.

Doing a manual copy of the build directory would be tricky - you'd have to manually change all the build parameters in the makefiles. Also, that might break cmake's dependency management, which works well with the double add_subdirectory.

ronald-cron-arm reviewed (changes requested):
To reduce the number of cmake variables, what about to just define a prefix that applies to all targets in the Mbed TLS project. That way we probably won't need the change specific to documentation.

RcColes commented:
Hi @ronald-cron-arm, could you check to see if these changes were what you were thinking of. I've tried to make the minimal change required to allow the prefix/suffix for targets, and have also included a minor bugfix where some cmake statements were using legacy syntax that was causing compatibility issues. The new syntax is supported since at least cmake version 3.0, which is as far back as their public documentation goes.

Summer-ARM commented on library/CMakeLists.txt:
```diff
@@ -183,13 +192,13 @@ if(USE_SHARED_MBEDTLS_LIBRARY)
     set_target_properties(mbedcrypto PROPERTIES VERSION 2.24.0 SOVERSION 5)
```
Hi Raef,
Good patch!
Suggest replacing mbedcrypto with ${mbedcrypto_target} like others.
￼

RcColes replied:
Done, thanks for spotting that

ronald-cron-arm commented on CMakeLists.txt:
```diff
@@ -273,7 +273,7 @@ if(ENABLE_PROGRAMS)
 endif()
 
-ADD_CUSTOM_TARGET(apidoc
+ADD_CUSTOM_TARGET(${target_prefix}apidoc${target_suffix}
```
So far we discussed about a prefix but not a suffix. Do you really need a suffix as well?
Otherwise as target_prefix is an Mbed TLS CMake global variable better to prefix it itself with "mbedtls" thus
"mbedtls_target_prefix" instead "target_prefix".

RcColes replied:
Done.

ronald-cron-arm reviewed (changes requested):
Hi, yes this is along the lines I was thinking of. Thanks. I have a few comments.

ronald-cron-arm reviewed (approved):
Thanks for the changes. LGTM now.

bensze01 reviewed (changes requested):
We should update the references to the libraries in the CMake files under `programs/` and `tests/` as well.

As is, I think the build will fail if you define both mbedtls_target_prefix, and ENABLE_TESTING or ENABLE_PROGRAMS

ronald-cron-arm commented:
Good point, we should probably have at least a cmake build in all.sh where mbedtls_target_prefix is set to not empty string value.

RcColes commented:
> Good point, we should probably have at least a cmake build in all.sh where mbedtls_target_prefix is set to not empty string value.

Do you think it would be better to be added to a current test or as a new one?

ronald-cron-arm reviewed (changes requested):
Doing a build to test just this would be a bit of a waste of CI resources to me.
Setting the MBEDTLS_TARGET_PREFIX to a non empty value in component_test_full_cmake_gcc_asan() component seems a good option to me.
Other things I overlooked:
. it seems it would be more aligned with the common usage to use upper case for the variable, thus MBEDTLS_TARGET_PREFIX.
. please document the variable at the begin of the root CMakeLists.txt. As part of this documentation, please explain the reason of this variable: something along the lines of what you wrote in the PR description.
. this needs a change log.

RcColes commented:
Think I've covered all requested changes.

ronald-cron-arm commented on ChangeLog.d/add_MBEDTLS_TARGET_PREFIX_to_cmake.txt:
```diff
@@ -0,0 +1,6 @@
+Features
+   * Add MBEDTLS_TARGET_PREFIX CMake variable, which is prefixed to the start of
+     the mbedtls, mbedcrypto, mbedx509 and apidoc CMake targets. This can be
+     used by external CMake projects that include this one to avoid CMake
```
"to avoid CMake variable name clashes". It is rather to avoid target name clashes,

ronald-cron-arm commented on ChangeLog.d/add_MBEDTLS_TARGET_PREFIX_to_cmake.txt:
```diff
@@ -0,0 +1,6 @@
+Features
+   * Add MBEDTLS_TARGET_PREFIX CMake variable, which is prefixed to the start of
```
Rather "which is prefixed to the mbedtls, mbedcrypto, mbedx509 and apidoc CMake target names." than "which is prefixed to the start of the mbedtls, mbedcrypto, mbedx509 and apidoc CMake targets." ?

RcColes replied:
Done

ronald-cron-arm commented on CMakeLists.txt:
```diff
@@ -14,6 +14,11 @@
 #     CMake files. It is related to ZLIB support which is planned to be removed.
 #     When the support is removed, the associated include_directories command
 #     will be removed as well as this note.
+#   + CMake targets are designed to be able to be altered by calling CMake in
```
This is not related to "Include directories" thus should be in another section. Thus something like:
` # - MBEDTLS_TARGET_PREFIX: ...`

Otherwise just document MBEDTLS_TARGET_PREFIX here not the "mbedtls_target", "mbedcrypto_target" ... "internal" variables.
Something similar to what you have done for the change log will be ok.

RcColes replied:
Done

bensze01 reviewed (approved):
Looks good to me.

bensze01 commented:
We should wait for the CI to pass at least once with the new changes to `all.sh` before merging.

ronald-cron-arm commented on CMakeLists.txt:
```diff
@@ -14,6 +14,11 @@
 #     CMake files. It is related to ZLIB support which is planned to be removed.
 #     When the support is removed, the associated include_directories command
 #     will be removed as well as this note.
+# - MBEDTLS_TARGET_PREFIX:
```
Nitpick: Add an empty line before and not need for a "sub-section" thus:

\#     will be removed as well as this note.
\#
\# - MBEDTLS_TARGET_PREFIX: CMake targets are designed ...

ronald-cron-arm commented on ChangeLog.d/add_MBEDTLS_TARGET_PREFIX_to_cmake.txt:
```diff
@@ -0,0 +1,6 @@
+Features
+   * Add MBEDTLS_TARGET_PREFIX CMake variable, which is prefixed to the mbedtls,
+     mbedcrypto, mbedx509 and apidoc CMake targets. This can be used by external
```
"target names" instead of just "targets".

ronald-cron-arm reviewed (changes requested):
Two minor remaining things.

ronald-cron-arm commented:
The test_cmake_as_subdirectory component of all.sh is falling in the CI:

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
FAILED: 1
test_cmake_as_subdirectory: build: cmake 'as-subdirectory' build: command make -> 2
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

RcColes commented:
> The test_cmake_as_subdirectory component of all.sh is falling in the CI:
> 
> !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
> FAILED: 1
> test_cmake_as_subdirectory: build: cmake 'as-subdirectory' build: command make -> 2
> !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Have you got a link to the failed CI run so I can take a look?

RcColes commented:
> The test_cmake_as_subdirectory component of all.sh is falling in the CI:
> 
> !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
> FAILED: 1
> test_cmake_as_subdirectory: build: cmake 'as-subdirectory' build: command make -> 2
> !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

I think this should now be fixed. I've also removed the test from that was added to all.sh and added it to the `as-subdirectory` test instead, as this seems to be the correct place to test this new feature as it is how it is intended to be used

ronald-cron-arm reviewed (approved):
Thanks for all the changes. The test is indeed in a better place now. This looks good to me. But for a CI infrastructure glitch the PR is passing the CI.

bensze01 reviewed (approved):
Looks good to me too.

gilles-peskine-arm commented:
This change doesn't break existing builds so I'm going to go ahead and merge it. However there's a risk that `MBEDTLS_TARGET_PREFIX` breaks, since it isn't tested. I've filed https://github.com/ARMmbed/mbedtls/issues/3832 as a wish to have a test.

bensze01 commented:
@gilles-peskine-arm `MBEDTLS_TARGET_PREFIX` _is_ tested. The cmake_subproject test [was modified](https://github.com/ARMmbed/mbedtls/pull/3407/files#diff-0c4cb1578f651514fd7632223957d5a044b11fdc1dbd536d9d709a5c7299941c) in this PR to also test this feature.
